### PR TITLE
feat(query-typegoose): Adds the ability to use discriminators

### LINKED ISSUE #1320: Currently the Nestjs-Typegoose package doesn't allow the use of discriminated entities.
**Is your feature request related to a problem? Please describe.**
Currently the Typegoose package doesn't allow for the use of discriminated entities. It is a great method to have schema inheritance. 

https://mongoosejs.com/docs/discriminators.html

### Have you read the [Contributing Guidelines](https://github.com/doug-martin/nestjs-query/blob/master/CONTRIBUTING.md)?

Yes

**Describe the solution you'd like**

I'd like to be able to use discriminated entities. 

**Describe alternatives you've considered**

No alternatives. 

**Additional context**

None
Scott

### PR DESCRIPTION
**Pull request by smolinari**
Fixes #1320

Scott

### CODE REVIEW

**Comment by smolinari** [comment]
@doug-martin - I see I actually can merge this PR myself, but I'd like your review please, as I'm still not very good with TypeScript and I'm basically cheating by looking at another library for the solution (sort of). If you'd like more detail on that, let me know. 

Scott

**Comment by smolinari** [comment]
@JohnMcInall - would you be willing and have the time to look over this PR? 

Scott

**Comment by doug-martin** [comment]
@smolinari I'm not super familiar with mongoose/typegoose and the more advanced features, but I'll take a look.

**Comment by smolinari** [comment]
@doug-martin - Ok. I'd appreciate any more experienced insights, even if it's not on the overall functionality, but more on the basic code level. If you say it's ok. I'll get it merged. 

Scott

**Comment by JohnMcInall** [comment]
@smolinari I shall take a look when I have a little time today.

**Comment by smolinari** [comment]
Awesome @JohnMcInall. I appreciate it. 

Scott

File: packages/query-typegoose/src/module.ts [hunk 1]
```diff
@@ -1,10 +1,11 @@
 import { DynamicModule } from '@nestjs/common';
 import { TypegooseModule } from 'nestjs-typegoose';
-import { Class } from '@nestjs-query/core';
+import { TypegooseClassWithOptions } from 'nestjs-typegoose/dist/typegoose-class.interface';
```
Comment (john-mcinall) [inline packages/query-typegoose/src/module.ts:3]: This import path shouldn't be coming from the dist folder, it should be pointing to the ts file.

File: packages/query-typegoose/src/providers.ts [hunk 2]
```diff
@@ -1,25 +1,56 @@
...
+
+const isTypegooseClass = (item: TypegooseInput): item is TypegooseClass => isClass(item);
+
+const isTypegooseClassWithOptions = (item: ClassOrDiscriminator): item is TypegooseClassWithOptions =>
+  isTypegooseClass(item.typegooseClass);
 
 AssemblerSerializer((obj: Document) => obj.toObject({ virtuals: true }))(Document);
-function createTypegooseQueryServiceProvider<Entity extends Base>(model: Class<unknown>): FactoryProvider {
+
+function convertToTypegooseClass(item: TypegooseInput): ClassOrDiscriminator | undefined {
```
[the remark this answers is not in the document]
Comment (smolinari) [reply packages/query-typegoose/src/providers.ts:21]: I've changed it to this:

```ts
function ensureProperInput(item: TypegooseInput): ClassOrDiscriminator | undefined {
  if (isTypegooseClass(item)) {
    return { typegooseClass: item };
  }
  if (isTypegooseClassWithOptions(item)) {
    return item;
  }
  return undefined;
}

function createTypegooseQueryServiceProvider<Entity extends Base>(
  model: TypegooseClass | TypegooseClassWithOptions,
): FactoryProvider {
  const inputModel = ensureProperInput(model);
  if (!inputModel) {
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    throw new Error(`Model definitions ${model} is incorrect.`);
  }
  const modelName = inputModel.typegooseClass?.name;

  return {
    provide: getQueryServiceToken({ name: modelName }),
    useFactory(ModelClass: ReturnModelType<new () => Entity>) {
      // initialize default serializer for documents, this is the type that mongoose returns from queries
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      AssemblerSerializer((obj: DocumentType<unknown>) => obj.toObject({ virtuals: true }))(ModelClass);

      return new TypegooseQueryService(ModelClass);
    },
    inject: [getModelToken(modelName)],
  };
}
```
Is that better?

Scott

[the remark this answers is not in the document]
Comment (JohnMcInall) [reply packages/query-typegoose/src/providers.ts:21]: Yeah that is better :) Just so if it is used outside of where it is declared you can understand what it is doing better :)

As long as tests are passing I don't see any issues with it from the quick glance I took over it

**Review by smolinari — COMMENTED** [review]
@JohnMcInall - Puh. Only had a "Resolve Discussion" button and I clicked it and our discussion disappeared. But, I got what you said. Thanks!!! Commit is coming. Just have to figure out the GPG stuff. Been a while since I've worked with Github and PRs. 

Scott

File: packages/query-typegoose/src/providers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (smolinari) [reply packages/query-typegoose/src/providers.ts:21]: Haha... for a second our discussion was gone. But then I replied in a normal comment and now it is back. What the heck? Oh well. Glad it is still here. :)

Scott

**Comment by smolinari** [comment]
@doug-martin - I've gone ahead and merged this. Hope that was ok.

Scott

**Pull request merged**